### DISCUSSION TITLE
Synthetic data for wakeword training

### DISCUSSION THREAD

**Discussion by StuartIanNaylor** [FutureProofHomes/Satellite1-Hardware#37] [Ideas]
https://github.com/StuartIanNaylor/wake_word_capture
Some back history is my own experience with various opensource wakeword is they are not particulary accurate and often prone to false positives as negatives. I have always called the oddly name Mycroft precise that because its results are far from consumer grade in the way microWakeword currently is. Its actually not the wakeWord model at least so with Mycroft precise but the manner of the dataset.

This is all from 3 years ago but just recently wather that a databse of ml-commons and real datasets I have used purely synthetic data and apart from 'Noise' this is 100% synthetic TTS wake word that in the above has a Tflite that you can try that I am surprised how well it works as this is still really a 'toy' dataset lacing in voices.

I have made an improvement in the above as near all seem to provide not much more than a binary classification of Wakeword, Unknownwords and Noise.
This has several problems with a lack of cross entropy so that the model overfits on training, but interms of features there is a huge class imbalance.

I have fixed this to a certain extent by adding further classes and the above example is 'Computer' via a CRNN from https://github.com/google-research/google-research/tree/master/kws_streaming

What I do is quite simple by 1st creating a language database 'English' and creating a sylable and phonemetables and count, the SqLite db is in the above repo.

'Unknown' then becomes all words with the same sylables as the wakeword, as an aproximation of key spectra in MFCC.
With a single word 2 classes 'LikeKW1', 'LikeKW2' that uses a phoneme slection to create 'Sounds like' on the front and last Sylables.
These words are excluded from Unknown because of the way softmax works but they make the training work harder to find distinguising features.
Then a '1syl' class of one sylable words to try and force overall edge/texture detection and one I call 'Phon' that is the complete KW duration of words concatenated and trimmed in a similar manner to noise again forcing overall edge/texture detection.
So end up with Kw, LikeKw1, LikeKw2, 1syl, Phon NotKw (unknown) and Noise as classifications.

Now I use TTS and a toy dataset of voices from the following.
Coqui ⓍTTSv2 870 https://github.com/coqui-ai/TTS
Emotivoice 1932 https://github.com/netease-youdao/EmotiVoice
Piper 904 https://github.com/k2-fsa/sherpa-onnx
Kokorov1 53 https://github.com/k2-fsa/sherpa-onnx
Kokorov1_1 103 https://github.com/k2-fsa/sherpa-onnx
Kokoro_en 11 https://github.com/k2-fsa/sherpa-onnx
VCKT 109 https://github.com/k2-fsa/sherpa-onnx

I say 'toy' as its just shy of 4000 wakeword with the 14000 of NotKw that sets my class sample size that all are augmented up to that qty.
TTS are great as its key to get clean samples so you can be accurate in augmentation of noise and reverberation even if I don't bother with reverb.
From phonetic columns various SqLite group by clauses, to try to create some even distribution of Phones and augmentation levels start with clean to assure end result. That Wakeword voice exists accross classification so that edges of spectra becomes more important than any possible textures that different recordings and different datasets can provide.

I have added resultant Tflite of a streaming / nonstreaming version and quantised version with the training logs and started with the basic Kw,NotKw ,Noise then add the x2 LikeKw, then the 1Syl and finally phon all in different training runs and included the logs so you can see the training curves this provides.

Keyword models work better with more classifications and your dataset design should try to force equal distribution of features whilst more means there is less chance of softmax triggering not because it has a high feature hit but that, hits are extremely low in all other classes.

It works well but I stopped at just shy of 4000 voices which obviously for all the prosody of input spoken english could have, is hugely overfiited. There is aproblem that modern TTS like much in opensource speech tech have language covered but is completely shy when it comes to dialect, accents and the varied prosody we meet in the wild.
It actually works pretty damn well but even using cloning TTS trying to find a source of dialect / accent datasets that have good metadata with good prosody range, seems a huge struggle.
Otherwise I would of continued and maybe moved on to providing Transfer learning so that I can create other wakeword with smaller datasets.

I did start off this quest 3 years ago with real data but that lack of clean recordings of any use that are not incorrectly labelled due to problems of forced alignment has stopped me using the like of CommonVoice and ML-Commons that also don't have metadata to filter and create balanced datasets.

I am pleasantly surprised by the results as was expecting less with what is purely an example toy dataset that seems far more tolerant to false positives, accurate to the KW and works at least out to 3meter without a microphone array or noise filter.
Classification models just work better with more classes to spread and balance feature and likely further can be add by adding similar sylable KW but unique phonetics and also gives choice of wakeword in use with no increase in model size and from what I can see no observable compute increase, yet at least. You can make a better wakwword model by adding more KW and maybe why Google just offered a few KW in a single model on offer, but they all add further classes all adding to the accuracy and stability of the model with no size increase or any noticeable increase in compute, apart from the added choice of multiple KW at the runtime.

I have a basic understanding of ML and classification wakeword models where with a simple model results are very dependent on the dataset.

Also anyone got any hinst on how to get more voices that are not just 'Neutral' english more representation of the many dialects and nations that use English as a 2nd language and as some of you might think the same of those in England the same :)
Anyone know of and large dialect prosody dataset of phonetic pangram sentences to use with one of the latest cloning TTS?
https://clagnut.com/blog/2380#English_phonetic_pangrams or any short sentances of dialects and accents of varied prosody with good metadata so you can create blalanced datasets.
You can not just pour a sample soup into a 'unknown' and expect consumer grade results.

As well as the tflite quant and unquantised models of a CRNN I added 2 trained models of https://github.com/Qualcomm-AI-research/bcresnet as its parameters start at 10k vs 400k for the CRNN but actually manages better results and is likely the SoTa of current wakeword and being Pytorch is a better fit for the active ONNX libs of https://github.com/espressif/esp-dl and also https://github.com/rockchip-linux/rknn-toolkit2 to run on NPU.
The datasets methods work with any classification model and if it was going to step up to production many more voices would need to be added and then that model would be used with transfer learning. This would allow much smaller datasets of different wakeword to be create but garner the accuracy of the parent model.

The repo has a link to the 'computer' dataset if you wish to download and examine.
